File: microservices/movie-service/src/main/java/com/project/movie/services/MovieService.java
package com.project.movie.services;

import com.project.movie.entities.Movie;

import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

public interface MovieService {
	
	public Flux<Movie> findAll();

	public Mono<Movie> findById(String movieId);
	
	public Flux<Movie> findByLanguage(String language);
	
	public Flux<Movie> findByGenre(String genre);
	
	public Mono<Movie> addMovie(Movie movie);
	
	public Mono<?> deleteMovie(String movieId);
	
	public Mono<Movie> updateMovie(Movie movie);
}

File: ui/movie-booking-app/src/app/services/movie.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { Movie } from '../models/movie.model';

@Injectable({
  providedIn: 'root'
})
export class MovieService {
  
  _baseUrl = `${environment.baseUrl}/movie-service/movie`;

  constructor( private http: HttpClient ) { }

  getAllMovies(): Observable<Movie[]> {
    return this.http.get<Movie[]>(`${this._baseUrl}`);
  }
  
  getMovieById(movieId: string): Observable<Movie> {
    return this.http.get<Movie>(`${this._baseUrl}/${movieId}`);
  }

  getMoviesByLanguage(language: string): Observable<Movie[]> {
    return this.http.get<Movie[]>(`${this._baseUrl}/language/${language}`);
  }

  getMoviesByGenre(genre: string): Observable<Movie[]> {
    return this.http.get<Movie[]>(`${this._baseUrl}/genre/${genre}`);
  }

  addMovie(movie: Movie): Observable<Movie> {
    return this.http.post<Movie>(`${this._baseUrl}`, movie);
  }

  updateMovie(movie: Movie): Observable<Movie> {
    return this.http.put<Movie>(`${this._baseUrl}`, movie);
  }

  deleteMovie(movieId: string): Observable<any> {
    return this.http.delete<any>(`${this._baseUrl}/${movieId}`);
  }
}
